basic_autonomy/src/helper_functions.cpp

```diff
@@ -64,20 +65,28 @@ namespace waypoint_generation
 
     int get_nearest_index_by_downtrack(const std::vector<lanelet::BasicPoint2d>& points, const carma_wm::WorldModelConstPtr& wm, double target_downtrack)
     {
-        size_t best_index = points.size() - 1;
-        for(size_t i = 0;i < points.size(); i++){
-            double downtrack = wm->routeTrackPos(points[i]).downtrack;
-            if(downtrack > target_downtrack){
-                //If value is negative, best index should be index 0
-                best_index = std::max((size_t)0, i - 1);
+        if(points.empty()){
+            RCLCPP_WARN_STREAM(rclcpp::get_logger(BASIC_AUTONOMY_LOGGER), "Empty points vector received, returning -1");
+            return -1;
+        }
 
-                RCLCPP_DEBUG_STREAM(rclcpp::get_logger(BASIC_AUTONOMY_LOGGER), "get_nearest_index_by_downtrack>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y() << ", downtrack: "<< downtrack);
-                break;
-            }
+        // Find first point with a downtrack greater than target_downtrack
+        const auto itr = std::find_if(std::cbegin(points), std::cend(points), 
+            [&wm = std::as_const(wm), target_downtrack](const auto & point) { return wm->routeTrackPos(point).downtrack > target_downtrack; });
+
+        int best_index = points.size() - 1;
```

Comment:
Optional: Consider using `std::size(points) - 1` instead of the member function. This aligns with the free function use for `std::cbegin()` and `std::cend()`.

Reply:
Good catch, just updated